include/wil/registry_basic.h

```diff
@@ -627,40 +354,56 @@ namespace wil
                 {
                     string.resize(byteSize / sizeof(wchar_t));
                 }
-                CATCH_LOG()
-                    return true;
+                catch (...)
+                {
+                    return false;
+                }
+                return true;
             }
 #endif // #if defined(_STRING_) && defined(WIL_ENABLE_EXCEPTIONS)
 
 #if defined(__WIL_OLEAUTO_H_)
+            inline bool grow_buffer_if_supported(BSTR& string, DWORD byteSize) WI_NOEXCEPT
+            {
+                // convert back to length (number of WCHARs)
+                DWORD length = byteSize / sizeof(WCHAR);
+                // SysAllocStringLen adds a null, so subtract a wchar_t from the input byteSize
+                length = length > 0 ? length - 1 : length;
+                const BSTR newString{ ::SysAllocStringLen(string, length) };
+                if (!newString)
+                {
+                    return false;
+                }
+                string = newString;
+                return true;
+            }
             inline bool grow_buffer_if_supported(::wil::unique_bstr& string, DWORD byteSize) WI_NOEXCEPT
             {
-                if (const BSTR newString{ ::SysAllocStringByteLen(nullptr, byteSize) })
+                BSTR newBstr = string.get();
+                if (::wil::reg::reg_view_details::grow_buffer_if_supported(newBstr, byteSize))
                 {
-                    ::memset(newString, 0, byteSize);
-                    const auto originalStringByteSize = ::SysStringLen(string.get()) * sizeof(WCHAR);
-                    const auto bytesToCopy = originalStringByteSize < byteSize ? originalStringByteSize : byteSize;
-                    ::memcpy(newString, string.get(), bytesToCopy);
-                    string.reset(newString);
+                    string.reset(newBstr);
+                    return true;
                 }
-                return true;
+                return false;
             }
 #endif // #if defined(__WIL_OLEAUTO_H_)
 
-#if defined(__WIL_OLEAUTO_H_STL)
-            inline bool grow_buffer_if_supported(::wil::shared_bstr& string, DWORD byteSize) WI_NOEXCEPT
+#if defined(__WIL_OBJBASE_H_)
+            inline bool grow_buffer_if_supported(::wil::unique_cotaskmem_string& string, DWORD byteSize) WI_NOEXCEPT
             {
-                if (const BSTR newString{ ::SysAllocStringByteLen(nullptr, byteSize) })
+                size_t length = byteSize / sizeof(wchar_t);
+                // subtracting 1 wchar_t in length because ::wil::make_unique_string_nothrow adds one to the length when it allocates

```

Comment:
We have to add/subtract 1 all over the place---should we instead change what number we use for the _input_ size, so it excludes the null ptr?

Reply:
I have to subtract 1 before calling make_unique_string_nothrow. I put this on its own line so I could comment as to why I'm going it vs. just putting it inline to the call to make_unique_string_nothrow.
These are functions for generic template code, so it must account for the unique behavior of whatever types/whatever functions we are using.
I preferred to just use standard wil string functions, and deal with the +/- length logic, vs. implementing my own.

Reply:
It seems like we've simplified this section of code a good deal. Closing.